src/main/webapp/resources/js/pages/projects/samples-metadata-import/services/importReducer.ts

```diff
@@ -0,0 +1,251 @@
+import { ImportDispatch, ImportState } from "../store";
+
+interface MetadataValidateDetailsItem {
+  isSampleNameValid: boolean;
+  foundSampleId?: number;
+}
+
+interface MetadataSaveDetailsItem {
+  saved: boolean;
+  error?: string;
+}
+interface SaveMetadataResponse {
+  metadataSaveDetails: Record<string, MetadataSaveDetailsItem>;
+}
+
+interface SetSampleNameColumnResponse {
+  sampleNameColumn: string;
+  metadataValidateDetails: Record<string, MetadataValidateDetailsItem>;
+}
+
+export interface InitialState {
+  sampleNameColumn: string;
+  headers: string[];
+  metadata: MetadataItem[];
+  metadataValidateDetails: Record<string, MetadataValidateDetailsItem>;
+  metadataSaveDetails: Record<string, MetadataSaveDetailsItem>;
+}
+
+const initialState: InitialState = {
+  sampleNameColumn: "",
+  headers: [],
+  metadata: [],
+  metadataValidateDetails: {},
+  metadataSaveDetails: {},
+};
+
+/*
+Redux async thunk for saving the metadata to samples.
+For more information on redux async thunks see: https://redux-toolkit.js.org/api/createAsyncThunk
+ */
+export const saveMetadata = createAsyncThunk<
+  SaveMetadataResponse,
+  { projectId: string; selectedMetadataKeys: string[] },
+  { dispatch: ImportDispatch; state: ImportState }
+>(
+  `importReducer/saveMetadata`,
+  async ({ projectId, selectedMetadataKeys }, { dispatch, getState }) => {
+    const state: ImportState = getState();
+    const sampleNameColumn: string = state.importReducer.sampleNameColumn;
+    const headers: string[] = state.importReducer.headers;
+    const metadata: MetadataItem[] = state.importReducer.metadata;
+    const metadataValidateDetails: Record<string, MetadataValidateDetailsItem> =
+      state.importReducer.metadataValidateDetails;
+    const metadataSaveDetails: Record<string, MetadataSaveDetailsItem> = {};
+
+    const chunkSize: number = 100;
+    for (let i = 0; i < metadata.length; i = i + chunkSize) {
+      const promises: Promise<any>[] = [];
+      for (let j = i; j < i + chunkSize && j < metadata.length; j++) {
+        const metadataItem: MetadataItem = metadata[j];
+        const index: string = metadataItem.rowKey;
+        if (
+          selectedMetadataKeys.includes(index) &&
+          metadataSaveDetails[index]?.saved !== true
+        ) {
+          const name: string = metadataItem[sampleNameColumn];
+          const metadataFields: FieldUpdate[] = Object.entries(metadataItem)
+            .filter(
+              ([key, value]) =>
+                headers.includes(key) && key !== sampleNameColumn
+            )
+            .map(([key, value]) => ({ field: key, value }));
+          const sampleId = metadataValidateDetails[index].foundSampleId;
+          if (sampleId) {
+            promises.push(
+              updateSample({
+                projectId,
+                sampleId,
+                body: {
+                  name,
+                  // TODO: Don't overwrite organism & description
```

Comment:
Are you updating this in a future branch?

Reply:
It's a reminder for me to circle back. I'm not sure what to do here. This endpoints updates the organism & description to whatever is past in the request. If it's null or empty, it's going to wash the current values. I only store the sampleId when checking if the sample exists. I don't pull all the sample details, not sure we want to. Since description can be empty, not sure i want to update the endpoint either, because maybe that's what the user wants to do. Any suggestions for me???